nurse_scheduling.py

```diff
@@ -0,0 +1,205 @@
+# Overall hard shift constraint:
+# lagrange_parameter * sum_d ((sum_n(effort * q_i(n,d)) - workforce) ** 2)
+#
+# with constant effort and constant workforce:
+# = lagrange_parameter * sum_d ( effort * sum_n q_i(n,d) - workforce ) ** 2
+# = lagrange_parameter * sum_d [ effort * effort * (sum_n q_i(n,d) ** 2)
+#                              - 2 effort * workforce * sum_n q_i(n,d)
+#                              + workforce * workforce ]
+# The constant term is moved to the offset, below, right before we solve
+# the QUBO
+#
+# Expanding and merging the terms:
+# lagrange_parameter * (effort * effort - 2 effort * workforce) *
+# sum_d sum_n q_i(n,d)
+# + lagrange_parameter * effort * effort * sum_d sum_m sum_n q_i(n,d) q_j(m, d) #
+for nurse_day_1 in range(size):
+    _, date_index = get_nurse_and_day(nurse_day_1)
+    # Diagonal term, without the workforce * workforce
+    Q[nurse_day_1, nurse_day_1] += lagrange_parameter * ((effort * effort) - (2 * workforce * effort))
+    for nurse_day_2 in range(size):
+        _, day_index_2 = get_nurse_and_day(nurse_day_2)
+        # Include only the same day, across nurses
+        if (date_index == day_index_2 and nurse_day_2 != nurse_day_1):
+            Q[nurse_day_1, nurse_day_2] += 2 * lagrange_parameter * effort * effort
+
+# Soft nurse constraint: all nurses should have approximately even work
+#                        schedules
+# This constraint tries to make preference * sum(q_i) equal to min_duty_days,
+# so that the nurses have the same number of days. The sum of the q_i,
+# over the number of days, is each nurse's number of days worked in the
+# schedule.
+# Overall soft nurse constraint:
+# gamma * sum_n ((sum_d(preference * q_i(n,d)) - min_duty_days) ** 2)
+# with constant preference and constant min_duty_days:
+# = gamma * sum_n ( preference * sum_d q_i(n,d) - min_duty_days ) ** 2
+# = gamma * sum_n [ preference * preference * (sum_d q_i(n,d) ** 2)
+#                              - 2 preference * min_duty_days * sum_d q_i(n,d)
+#                              + min_duty_days * min_duty_days ]
+# The constant term is moved to the offset, below, right before we solve
+# the QUBO
+#
+# The square of the the sum_d term becomes:
+# Expanding and merging the terms:
+# = gamma * (preference * preference - 2 preference * min_duty_days) * sum_n sum_d q_i(n,d)
+# + gamma * preference * preference * sum_n sum_d1 sum_d2 q_i(n,d1)
+#                      * q_j(n, d2)
+
+for nurse_day_1 in range(size):
+    nurse_index_1, _ = get_nurse_and_day(nurse_day_1)
+    # Diagonal term, without the min_duty_days * min_duty_days
+    Q[nurse_day_1, nurse_day_1] += gamma * ((preference * preference) - (2 * min_duty_days * preference))
+    for nurse_day_2 in range(nurse_day_1 + 1, size):
+        nurse_index_2, _ = get_nurse_and_day(nurse_day_2)
+        if (nurse_index_1 == nurse_index_2 and nurse_day_2 != nurse_day_1):
+            Q[nurse_day_1, nurse_day_2] += 2 * gamma * preference * preference
```

Comment:
Similar to your other constraint, I believe this part of the code could benefit from being broken into two different parts - code that deals with the diagonal and off-diagonal terms separately. Also, looping through the nurse and day indices directly and letting your composite-index helper deal with the composite indices will make the code easier to read.

Reply:
Yes indeed. Thanks. Please have a look.